include/exadg/operators/operator_base.cpp

```diff
@@ -262,45 +260,34 @@ OperatorBase<dim, Number, n_components>::apply(VectorType & dst, VectorType cons
   if(is_dg)
   {
     if(evaluate_face_integrals())
+    {
       matrix_free->loop(&This::cell_loop,
                         &This::face_loop,
                         &This::boundary_face_loop_hom_operator,
                         this,
                         dst,
                         src,
                         true);
+    }
     else
+    {
       matrix_free->cell_loop(&This::cell_loop, this, dst, src, true);
+    }
   }
   else
   {
-    // We need to zero constrained degrees of freedom. The contribution of these
-    // degrees of freedom to the matrix-vector product has already been taken into account
-    // in the right-hand side vector. Put differently, setting the constrained degrees of
-    // freedom to zero is equivalent to the idea of eliminating those columns of the matrix
-    // that correspond to constrained degrees of freedom.
-    for(unsigned int i = 0; i < constrained_indices.size(); ++i)
-    {
-      constrained_values_src[i] = src.local_element(constrained_indices[i]);
-      const_cast<VectorType &>(src).local_element(constrained_indices[i]) = 0.;
-    }
-
-    // Compute matrix-vector product. Constrained degrees of freedom will not have
-    // an effect since they have been zeroed before.
+    // Compute matrix-vector product. Constrained degrees of freedom in the src-vector will not be
+    // used. The function read_dof_values() (or gather_evaluate()) uses the homogeneous boundary
+    // data passed to MatrixFree via AffineConstraints.
     matrix_free->cell_loop(&This::cell_loop, this, dst, src, true);
 
     // Constrained degree of freedom are not removed from the system of equations.
     // Instead, we set the diagonal entries of the matrix to 1 for these constrained
     // degrees of freedom. This means that we simply copy the constrained values to the
-    // dst vector. Since we have already computed the matrix-vector product, the
-    // constrained values can be written back to the src vector, i.e. having completed
-    // the function apply(), the src vector should be the same as before calling this
-    // function.
+    // dst vector.
     for(unsigned int i = 0; i < constrained_indices.size(); ++i)
     {
-      const_cast<VectorType &>(src).local_element(constrained_indices[i]) =
-        constrained_values_src[i];
-      dst.local_element(constrained_indices[i]) = constrained_values_src[i];
+      dst.local_element(constrained_indices[i]) = src.local_element(constrained_indices[i]);
```

Comment:
Do we still need this? Since we fill `constrained_indices` by `this->matrix_free->get_constrained_dofs(this->data.dof_index)`, we could also just do it here:
```c++
  for (const unsigned int constrained_index : matrix_free->get_constrained_dofs(this->data.dof_index)
    dst.local_element(constrained_index) = src.local_element(constrained_index);
```

Reply:
The function returns exactly a reference to `std::vector<unsigned int>`, https://github.com/dealii/dealii/blob/c9b6a28e288d5c9d955ee5a3befd181bcda3b8d9/include/deal.II/matrix_free/matrix_free.h#L2432-L2438 - so the underlying cost of the range-based for loop should be exactly the same (apart from one or two additional indirection per outer call to `vmult` to follow into the variables of `matrix_free`, which is essentially nothing in the grand scheme of things).